docs/ocis/adr/0023-dynamic-app-registry.md

```diff
@@ -0,0 +1,51 @@
+* Status: proposed
+* Deciders: @butonic, @dragonchaser
+* Date: 2023-09-28
+
+## Context and Problem Statement
+
+The intent of the CS3 API is to link enterprise file share and sync platforms with storage and application providers. The CS3 [RegistryAPI](https://cs3org.github.io/cs3apis/#cs3.app.registry.v1beta1.RegistryAPI) allows registering CS3 [AppProviders](https://cs3org.github.io/cs3apis/#cs3.app.registry.v1beta1.ProviderInfo). However, the API does not explain if and how often an AppProvider needs to reregister. In K8s ip addresses may change anytime and pods can get started and removed at any time. Without a mechanism to reupdate the App Registry when that happens the metadata of an app provider - the filetype it can handle and the endpoint to use - will be invalid or forgotten.
+
+## Out of scope
+Additional app properties like web UI elements or scripts that Web could use to dynamically load extensions. That will be part of a seperate ADR.
+
+## Decision Drivers
+
+## Considered Options
+
+* [Dynamically reregister App Providers](#dynamic-registry)
+* [Switch to a Go Micro based Registry](#micro-registry)
+
+## Decision Outcome
+
+Chosen option: [Dynamically reregister App Providers](#dynamic-registry)
+
+### Positive Consequences
+
+* App-providers can register service endpoints, supported mimetypes
+* App-providers gain the ability to refresh their data in the registry, the registry gains a ttl for each service resulting in automatic removal & cleanup of dead services.
+* CS3 API usage clarified and we are not bypassing by reading the app provider metadata from the go micro registry.
+
+### Negative Consequences
+
+* We need to update the CS3 API spec and add a TTL to the [ProviderInfo](https://cs3org.github.io/cs3apis/#cs3.app.registry.v1beta1.ProviderInfo) so it can can be used in the [ListAppProvidersResponse](https://cs3org.github.io/cs3apis/#cs3.app.registry.v1beta1.ListAppProvidersResponse) but more importantly in the [AddAppProviderRequest](https://cs3org.github.io/cs3apis/#cs3.app.registry.v1beta1.AddAppProviderRequest) so App providers can optionally request a TTL. Finally, the [AddAppProviderResponse](https://cs3org.github.io/cs3apis/#cs3.app.registry.v1beta1.AddAppProviderResponse) also needs to return a TTL so the App Provider knows when to reregister.
+* The static app registry implementation needs a config option for a TTL
+
+## Pros and Cons of the Options
+
+### Dynamically reregister App Providers
+The App Registry would use a TTL to determine when to forget App Providers and App Providers would have to refresh their registration periodically.
```

Comment:
Tell clients the TTL via cache header proposed by @fmoc in #5754 ?

Reply:
The purpose of the TTL is to remove app-providers from the registry that are no longer available (e.g. failed to refresh their entry in the registry, on the long run this should result in removal from the web-ui). I am not familiar with the QT-Client, does this directly access apps like onlyoffice or collabora? If not I do not understand why this is related to the client, can you please clarify that for me? Thx.

Reply:
HTTP clients (outside) should also know how often refresh apps/list. This was my association when reading TTL

Reply:
Valid point. Not sure if we should mention that in the ADR, but  it definitely belongs into the implementation.